// @valencets/resultkit@0.1.1/dist/result.js
import { ResultAsync } from './result-async.js';
export class Ok {
    value;
    constructor(value) {
        this.value = value;
    }
    isOk() {
        return true;
    }
    isErr() {
        return false;
    }
    map(fn) {
        return new Ok(fn(this.value));
    }
    mapErr(_fn) {
        return new Ok(this.value);
    }
    andThen(fn) {
        return fn(this.value);
    }
    orElse(_fn) {
        return new Ok(this.value);
    }
    match(okFn, _errFn) {
        return okFn(this.value);
    }
    unwrapOr(_defaultValue) {
        return this.value;
    }
    unwrap() {
        return this.value;
    }
    unwrapErr() {
        // eslint-disable-next-line no-restricted-syntax
        throw new Error('Called unwrapErr on an Ok value');
    }
    toAsync() {
        return new ResultAsync(Promise.resolve(this));
    }
}
export class Err {
    error;
    constructor(error) {
        this.error = error;
    }
    isOk() {
        return false;
    }
    isErr() {
        return true;
    }
    map(_fn) {
        return new Err(this.error);
    }
    mapErr(fn) {
        return new Err(fn(this.error));
    }
    andThen(_fn) {
        return new Err(this.error);
    }
    orElse(fn) {
        return fn(this.error);
    }
    match(_okFn, errFn) {
        return errFn(this.error);
    }
    unwrapOr(defaultValue) {
        return defaultValue;
    }
    unwrap() {
        // eslint-disable-next-line no-restricted-syntax
        throw new Error('Called unwrap on an Err value');
    }
    unwrapErr() {
        return this.error;
    }
    toAsync() {
        return new ResultAsync(Promise.resolve(this));
    }
}
export function ok(value) {
    return new Ok(value);
}
export function err(error) {
    return new Err(error);
}
