// @valencets/resultkit@0.1.1/dist/result-async.js
import { ok, err } from './result.js';
export class ResultAsync {
    _promise;
    constructor(promise) {
        this._promise = promise;
    }
    static fromPromise(promise, errorFn) {
        return new ResultAsync(Promise.resolve(promise).then((value) => ok(value), (e) => err(errorFn(e))));
    }
    static fromSafePromise(promise) {
        return new ResultAsync(Promise.resolve(promise).then((value) => ok(value)));
    }
    map(fn) {
        return new ResultAsync(this._promise.then(async (result) => {
            if (result.isErr()) {
                return err(result.error);
            }
            const mapped = fn(result.value);
            const resolved = mapped instanceof Promise ? await mapped : mapped;
            return ok(resolved);
        }));
    }
    mapErr(fn) {
        return new ResultAsync(this._promise.then(async (result) => {
            if (result.isOk()) {
                return ok(result.value);
            }
            const mapped = fn(result.error);
            const resolved = mapped instanceof Promise ? await mapped : mapped;
            return err(resolved);
        }));
    }
    andThen(fn) {
        return new ResultAsync(this._promise.then((result) => {
            if (result.isErr()) {
                return err(result.error);
            }
            const next = fn(result.value);
            if (next instanceof ResultAsync) {
                return next._promise;
            }
            return next;
        }));
    }
    orElse(fn) {
        return new ResultAsync(this._promise.then((result) => {
            if (result.isOk()) {
                return ok(result.value);
            }
            const next = fn(result.error);
            if (next instanceof ResultAsync) {
                return next._promise;
            }
            return next;
        }));
    }
    async match(okFn, errFn) {
        const result = await this._promise;
        if (result.isOk()) {
            return okFn(result.value);
        }
        return errFn(result.error);
    }
    async unwrapOr(defaultValue) {
        const result = await this._promise;
        if (result.isOk()) {
            return result.value;
        }
        return defaultValue;
    }
    then(onFulfilled, onRejected) {
        return this._promise.then(onFulfilled, onRejected);
    }
}
export function okAsync(value) {
    return new ResultAsync(Promise.resolve(ok(value)));
}
export function errAsync(error) {
    return new ResultAsync(Promise.resolve(err(error)));
}
